elm/diff_model.py

```diff
@@ -70,12 +75,153 @@ def unsafe_execute(code_str: str, timeout: int = 5):
+        Returns:
+            a list of code(s) generated by the model.
+        """
+        tokenizer = self.tokenizer if tokenizer is None else tokenizer
+        encoding = tokenizer(
+            [seed + '\n\n' + self.func_preamble],
+            truncation=True,
+            padding=True,
+            max_length=self.cfg.gen_max_len,
+            return_tensors="pt",
+        )
+
+        cfg = OmegaConf.merge(self.cfg, {"batch_size": self.cfg.batch_size if batch_size is None else batch_size})
+        with torch.no_grad():
+            completion = sample(cfg, self.model, self.tokenizer, encoding)
+        # Reset random seed
+        set_seed(int(self.rng.integers(0, 1e8)))
+
+        if without_trunc:
+            truncation = completion
+        else:
+            truncation = [truncate(code, print_num=float('inf'), only_local_scope=True) for code in completion]
+
+        truncation = [self.import_line + '\n' + self.func_preamble + '\n' + code for code in truncation]
+
+        if append_return:
+            truncation = [code + '\n' + self.return_line for code in truncation]
+
+        return truncation
+
+    def generate_program(self, code: str) -> list[dict]:
+        """
+        Given a piece of code, do prompt mutation, call the sandbox server to execute the code and return the result.
+        Args:
+            code: the full code string.
+        Returns:
+            a numpy array (if successful) or the exception object.
+        """
+        results = []
+        for code in self.generate_prompt_str(code):
+            resp = self._get_response(code, self.cfg.timeout)
+            if resp.status_code == 200:
+                return_dict = json.loads(resp.text)
+                self._post_process(return_dict)
+                error_code = "0"
+            elif resp.status_code == 500:  # Bad request
+                try:
+                    msg = json.loads(resp.text)
+                    return_dict = {"program_str": code, "result_obj": msg["message"]}
+                    error_code = msg["unsafe_execute_error_code"]
+                except Exception as e:
+                    return_dict = {"program_str": code, "result_obj": str(e)}
+                    error_code = 6
+            else:
+                return_dict = {"program_str": code, "result_obj": resp.text}
+                error_code = 6
+
+            results.append({**return_dict, "error_code": error_code})
+
+        return results
+
+    @abstractmethod
+    def _get_response(self, code: str, timeout: int) -> requests.models.Response:
+        pass
+
+    @abstractmethod
+    def _post_process(self, response_dict: dict) -> dict:
+        pass
+
+
+class PromptMutationForSodarace(PromptMutationModel):
+    func_name: str = "make_walker"
+    import_line: str = "from .walker import walker_creator"
+    func_preamble: str = f"def {func_name}():\n\twc = walker_creator()\n"
+    return_line: str = "\treturn wc.get_walker()\n"
```

Comment:
These properties don't seem to be used in this class. Is this on purpose?

Reply:
Actually they are (in the parent abstract class lol). I assume the way to construct prompts are the same procedure. This cuts down some duplications of codes but yeah, it also causes misunderstandings.

Reply:
see line 116-141